File: c/algorithms/allocation/first_fit.c
/*
 * first_fit.c
 * First Fit allocation: scan from index 0 and allocate the first
 * free block >= requested size. O(n) per allocation.
 * Fast but causes clustering of small fragments at low addresses.
 * Time complexity: O(n) allocation, O(n) free + coalesce.
 */

#include "first_fit.h"
#include <string.h>

static void coalesce_free_blocks(MemoryMap *map) {
    int i;
    for (i = 0; i < map->block_count - 1; ) {
        if (map->blocks[i].is_free && map->blocks[i + 1].is_free) {
            map->blocks[i].size += map->blocks[i + 1].size;
            int j;
            for (j = i + 1; j < map->block_count - 1; j++) {
                map->blocks[j] = map->blocks[j + 1];
            }
            map->block_count--;
        } else {
            i++;
        }
    }
}

static void split_block(MemoryMap *map, int index, int size) {
    if (map->blocks[index].size == size) return;
    if (map->block_count >= MAX_BLOCKS) return;
    int j;
    for (j = map->block_count; j > index + 1; j--) {
        map->blocks[j] = map->blocks[j - 1];
    }
    map->blocks[index + 1].id = map->block_count;
    map->blocks[index + 1].base = map->blocks[index].base + size;
    map->blocks[index + 1].size = map->blocks[index].size - size;
    map->blocks[index + 1].is_free = 1;
    map->blocks[index + 1].process_id = -1;
    map->blocks[index].size = size;
    map->block_count++;
}

int first_fit_allocate(MemoryMap *map, int process_id, int size) {
    int m = map->block_count;
    int n = 1;
    int blockSize[MAX_BLOCKS];
    int k;
    for (k = 0; k < m; k++) {
        blockSize[k] = map->blocks[k].is_free ? map->blocks[k].size : 0;
    }
    int processSize[1] = {size};
    int allocation[1] = {-1};

    int i = 0;
    int j;
    for (j = 0; j < m; j++) {
        if (blockSize[j] >= processSize[i]) {
            allocation[i] = j;
            blockSize[j] -= processSize[i];
            break;
        }
    }

    if (allocation[0] != -1) {
        j = allocation[0];
        split_block(map, j, size);
        map->blocks[j].is_free = 0;
        map->blocks[j].process_id = process_id;
        map->last_allocated_index = j;
        return map->blocks[j].base;
    }
    return -1;
}

void first_fit_free(MemoryMap *map, int process_id) {
    int i;
    for (i = 0; i < map->block_count; i++) {
        if (!map->blocks[i].is_free && map->blocks[i].process_id == process_id) {
            map->blocks[i].is_free = 1;
            map->blocks[i].process_id = -1;
        }
    }
    coalesce_free_blocks(map);
}

MemoryStats first_fit_stats(MemoryMap *map) {
    MemoryStats s;
    memset(&s, 0, sizeof(MemoryStats));
    s.total_size = map->total_size;
    s.block_count = map->block_count;
    int i;
    for (i = 0; i < map->block_count; i++) {
        if (map->blocks[i].is_free) {
            s.free_size += map->blocks[i].size;
            s.free_block_count++;
            if (map->blocks[i].size > s.external_fragmentation)
                s.external_fragmentation = map->blocks[i].size;
        } else {
            s.used_size += map->blocks[i].size;
        }
    }
    if (s.free_block_count > 0)
        s.external_fragmentation = s.free_size - s.external_fragmentation;
    return s;
}

File: js/algorithms/allocation/first_fit.js
/*
 * first_fit.js
 * First Fit allocation: scan from index 0 and allocate the first
 * free block >= requested size. O(n) per allocation.
 * Fast but causes clustering of small fragments at low addresses.
 * JS port of c/algorithms/allocation/first_fit.c
 */

function coalesce(blocks) {
  let i = 0;
  while (i < blocks.length - 1) {
    if (blocks[i].isFree && blocks[i + 1].isFree) {
      blocks[i].size += blocks[i + 1].size;
      blocks.splice(i + 1, 1);
    } else {
      i++;
    }
  }
}

function splitBlock(blocks, index, size) {
  const blk = blocks[index];
  if (blk.size === size) return;
  const remainder = {
    id: blocks.length,
    base: blk.base + size,
    size: blk.size - size,
    isFree: true,
    processId: -1
  };
  blk.size = size;
  blocks.splice(index + 1, 0, remainder);
}

export function makeMemoryMap(totalSize) {
  return {
    blocks: [{ id: 0, base: 0, size: totalSize, isFree: true, processId: -1 }],
    totalSize,
    lastAllocatedIndex: 0
  };
}

export function firstFitAllocate(map, processId, size) {
  let m = map.blocks.length;
  let n = 1;
  let blockSize = [];
  for(let k = 0; k < m; k++) {
    blockSize[k] = map.blocks[k].isFree ? map.blocks[k].size : 0;
  }
  let processSize = [size];
  let allocation = [-1];

  let i = 0;
  for(let j = 0; j < m; j++) {
      if(blockSize[j] >= processSize[i]) {
          allocation[i] = j;
          blockSize[j] -= processSize[i];
          break;
      }
  }

  if(allocation[0] !== -1) {
      let j = allocation[0];
      splitBlock(map.blocks, j, size);
      map.blocks[j].isFree = false;
      map.blocks[j].processId = processId;
      map.lastAllocatedIndex = j;
      return map.blocks[j].base;
  }
  return -1;
}

export function firstFitFree(map, processId) {
  for (let i = 0; i < map.blocks.length; i++) {
    if (!map.blocks[i].isFree && map.blocks[i].processId === processId) {
      map.blocks[i].isFree = true;
      map.blocks[i].processId = -1;
    }
  }
  coalesce(map.blocks);
}

export function firstFitStats(map) {
  return computeStats(map);
}

export function computeStats(map) {
  let usedSize = 0, freeSize = 0, freeBlockCount = 0, largestFree = 0;
  for (const b of map.blocks) {
    if (b.isFree) {
      freeSize += b.size;
      freeBlockCount++;
      if (b.size > largestFree) largestFree = b.size;
    } else {
      usedSize += b.size;
    }
  }
  const externalFrag = freeBlockCount > 1 ? freeSize - largestFree : 0;
  return {
    totalSize: map.totalSize,
    usedSize,
    freeSize,
    externalFragmentation: externalFrag,
    internalFragmentation: 0,
    blockCount: map.blocks.length,
    freeBlockCount
  };
}
